java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java

```diff
@@ -0,0 +1,152 @@
+  private boolean consumerCreated = false;
+
+  private VectorSchemaRoot nextBatch;
+
+  private final int partialLimit;
+
+  /**
+   * Construct an instance.
+   */
+  private ArrowVectorIterator(ResultSet resultSet, JdbcToArrowConfig config) throws SQLException {
+    this.resultSet = resultSet;
+    this.config = config;
+    this.schema = JdbcToArrowUtils.jdbcToArrowSchema(resultSet.getMetaData(), config);
+    this.partialLimit = config.getPartialLimit();
+
+    rsmd = resultSet.getMetaData();
+    consumers = new JdbcConsumer[rsmd.getColumnCount()];
+    this.compositeConsumer = new CompositeJdbcConsumer(consumers);
+  }
+
+  /**
+   * Create a ArrowVectorIterator to partially convert data.
+   */
+  public static ArrowVectorIterator create(
+      ResultSet resultSet,
+      JdbcToArrowConfig config)
+      throws SQLException {
+
+    ArrowVectorIterator iterator = new ArrowVectorIterator(resultSet, config);
+    try {
+      iterator.load();
+      return iterator;
+    } catch (Exception e) {
+      iterator.close();
+      throw new RuntimeException("error occurs while creating iterator", e);
+    }
+  }
+
+  // Loads the next schema root or null if no more rows are available.
+  private void load() throws IOException, SQLException {
+    VectorSchemaRoot root = VectorSchemaRoot.create(schema, config.getAllocator());
+    JdbcToArrowUtils.allocateVectors(root, JdbcToArrowUtils.DEFAULT_BUFFER_SIZE);
+
+    if (!consumerCreated) {
+      consumerCreated = true;
+      for (int i = 1; i <= consumers.length; i++) {
+        consumers[i - 1] = JdbcToArrowUtils.getConsumer(resultSet, i, resultSet.getMetaData().getColumnType(i),
+            root.getVector(rsmd.getColumnName(i)), config);
+      }
+    } else {
+      for (int i = 1; i <= consumers.length; i++) {
+        consumers[i - 1].resetValueVector(root.getVector(rsmd.getColumnName(i)));
+      }
+    }
+
+    // consume data
+    try {
+      int readRowCount = 0;
+      while (resultSet.next()) {
+        compositeConsumer.consume(resultSet);
+        readRowCount++;
+        if (partialLimit != JdbcToArrowConfig.DEFAULT_PARTIAL_LIMIT && readRowCount >= partialLimit) {
+          break;
+        }
+      }
+      root.setRowCount(readRowCount);
+    } catch (Exception e) {
+      compositeConsumer.close();
+      throw new RuntimeException("error occurs while consuming data.", e);
+    }
+
+    if (root.getRowCount() == 0) {
+      root.close();
+      nextBatch = null;
+    } else {
+      nextBatch = root;
+    }
+  }
+
+  @Override
+  public boolean hasNext() {
+    return nextBatch != null;
+  }
+
+  /**
+   * Gets the next vector. The user is responsible for freeing its resource from close.
+   */
+  public VectorSchemaRoot next() {
+    Preconditions.checkArgument(hasNext());
+    VectorSchemaRoot returned = nextBatch;
+    try {
+      load();
+    } catch (Exception e) {
+      close();
+      throw new RuntimeException("error occurs while getting next schema root");
```

Comment:
pass along the root cause.
```suggestion
      throw new RuntimeException("Error occured while getting next schema root.", e);
```